test/cpp/end2end/message_allocator_end2end_test.cc

```diff
@@ -0,0 +1,405 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <thread>
+
+#include <google/protobuf/arena.h>
+
+#include <gtest/gtest.h>
+
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+#include <grpcpp/support/client_callback.h>
+#include <grpcpp/support/message_allocator.h>
+
+#include "src/core/lib/iomgr/iomgr.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/test_credentials_provider.h"
+
+// MAYBE_SKIP_TEST is a macro to determine if this particular test configuration
+// should be skipped based on a decision made at SetUp time. In particular, any
+// callback tests can only be run if the iomgr can run in the background or if
+// the transport is in-process.
+#define MAYBE_SKIP_TEST \
+  do {                  \
+    if (do_not_test_) { \
+      return;           \
+    }                   \
+  } while (0)
+
+namespace grpc {
+namespace testing {
+namespace {
+
+class CallbackTestServiceImpl
+    : public EchoTestService::ExperimentalCallbackService {
+ public:
+  explicit CallbackTestServiceImpl() {}
+
+  void SetFreeRequest() { free_request_ = true; }
+
+  void SetAllocatorMutator(
+      std::function<void(void* allocator_state, const EchoRequest* req,
+                         EchoResponse* resp)>
+          mutator) {
+    allocator_mutator_ = mutator;
+  }
+
+  void Echo(ServerContext* context, const EchoRequest* request,
+            EchoResponse* response,
+            experimental::ServerCallbackRpcController* controller) override {
+    response->set_message(request->message());
+    if (free_request_) {
+      controller->FreeRequest();
+    } else if (allocator_mutator_) {
+      allocator_mutator_(controller->GetAllocatorState(), request, response);
+    }
+    controller->Finish(Status::OK);
```

Comment:
We need to actually define when the request and response have their ownership returned to the allocator. Is the intention that it would be at the time of the Finish() call? Technically we have a problem for future expansion that we haven't dealt with yet because if we want to add incremental serialization this will be a stumbling block since we don't have an OnDone for unary calls. We can deal with this right now by switching the unary calls to use a reactor model with a proper OnDone. That said, I don't think it's a requirement to get this PR in.

Reply:
For the request/response, I think once we reach the point where explicit DeallocateMessages is called, the custom allocator can handle the deallocation. As for the library, it should only call DeallocateMessages after any serialization is done.

Reply:
Yes, that's true. We can adjust that later as we want. No issue here.